frame/metadata/src/lib.rs

```diff
@@ -300,6 +300,11 @@ pub enum StorageEntryType {
 		value: DecodeDifferentStr,
 		key2_hasher: StorageHasher,
 	},
+	NMap {
+		keys: DecodeDifferentArray<&'static str, StringBuf>,
+		hashers: DecodeDifferentArray<StorageHasher>,
+		value: DecodeDifferentStr,
+	},
```

Comment:
I think we should bump the metadata version. Because third-parties won't be able to make sense of this new variant.

Thus releasing RuntimeMetadataV13:
```diff
diff --git a/frame/metadata/src/lib.rs b/frame/metadata/src/lib.rs
index a502af0798..ba232a88f1 100644
--- a/frame/metadata/src/lib.rs
+++ b/frame/metadata/src/lib.rs
@@ -369,8 +369,10 @@ pub enum RuntimeMetadata {
        V10(RuntimeMetadataDeprecated),
        /// Version 11 for runtime metadata. No longer used.
        V11(RuntimeMetadataDeprecated),
-       /// Version 12 for runtime metadata.
-       V12(RuntimeMetadataV12),
+       /// Version 12 for runtime metadata. No longer used.
+       V12(RuntimeMetadataDeprecated),
+       /// Version 13 for runtime metadata.
+       V13(RuntimeMetadataV13),
 }
 
 /// Enum that should fail.
@@ -394,7 +396,7 @@ impl Decode for RuntimeMetadataDeprecated {
 /// The metadata of a runtime.
 #[derive(Eq, Encode, PartialEq, RuntimeDebug)]
 #[cfg_attr(feature = "std", derive(Decode, Serialize))]
-pub struct RuntimeMetadataV12 {
+pub struct RuntimeMetadataV13 {
        /// Metadata of all the modules.
        pub modules: DecodeDifferentArray<ModuleMetadata>,
        /// Metadata of the extrinsic.
@@ -402,7 +404,7 @@ pub struct RuntimeMetadataV12 {
 }
 
 /// The latest version of the metadata.
-pub type RuntimeMetadataLastVersion = RuntimeMetadataV12;
+pub type RuntimeMetadataLastVersion = RuntimeMetadataV13;
 
 /// All metadata about an runtime module.
 #[derive(Clone, PartialEq, Eq, Encode, RuntimeDebug)]
@@ -430,6 +432,6 @@ impl Into<sp_core::OpaqueMetadata> for RuntimeMetadataPrefixed {
 
 impl Into<RuntimeMetadataPrefixed> for RuntimeMetadataLastVersion {
        fn into(self) -> RuntimeMetadataPrefixed {
-               RuntimeMetadataPrefixed(META_RESERVED, RuntimeMetadata::V12(self))
+               RuntimeMetadataPrefixed(META_RESERVED, RuntimeMetadata::V13(self))
        }
 }
```

Also we could remove other variant as they can all be expressed as a NMap but I think it is better for compatibility to keep them (basically `Plain` would be a `NMap` with `keys` and `hashers` being array of length 0, `Map` would be a `NMap` with keys and arrays of length 1, etc..). cc @jacogr 

(Also it would make more sense to have something like `DecodeDifferentArray<(DecodeDifferentStr, StorageHasher)>` or something like this maybe, to ensure keys and hashers are of same length.)

Reply:
Ok. This is the first I have seen this. This is a major revamp on my side for everything storage related.

Reply:
It yes, this is correct bumping - with metadata you can just change at will - it needs a bump to avoid issue like we have in the v9-late iteration metadata.

Reply:
@jacogr we wont spring anything on you here. This should be unused in the runtime for now, and might only get used in assets in the near future for a triple map.

Reply:
That is fine, appreciate it. Do bear in mind that once merged, it does immediately make Substrate master non-supported (if any enum changes are made to the storage descriptors or the metadata version is bumped), even if unused in practice.